### PR TITLE
[radio] add radio coex control api

### PR DESCRIPTION
**Pull request by zhanglongxia**

### CODE REVIEW

File: examples/platforms/posix/radio.c [hunk 1]
```diff
@@ -147,6 +147,10 @@ static uint16_t     sShortAddressMatchTable[POSIX_MAX_SRC_MATCH_ENTRIES];
 static otExtAddress sExtAddressMatchTable[POSIX_MAX_SRC_MATCH_ENTRIES];
 static bool         sSrcMatchEnabled = false;
 
+#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE
```
Comment (bukepo) [inline examples/platforms/posix/radio.c]: `OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE` -> `OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE`

File: src/cli/cli.cpp [hunk 2]
```diff
@@ -787,35 +787,54 @@ void Interpreter::ProcessCoapSecure(int argc, char *argv[])
...
+        mServer->OutputFormat("    Grant Wait: %u\r\n", metrics.mNumRxGrantWait);
+        mServer->OutputFormat("    Grant Wait Activated: %u\r\n", metrics.mNumRxGrantWaitActivated);
+        mServer->OutputFormat("    Grant Wait Timeout: %u\r\n", metrics.mNumRxGrantWaitTimeout);
+        mServer->OutputFormat("    Grant Deactivated During Request: %u\r\n",
+                              metrics.mNumRxGrantDeactivatedDuringRequest);
+        mServer->OutputFormat("    Delayed Grant: %u\r\n", metrics.mNumRxDelayedGrant);
+        mServer->OutputFormat("    Average Request To Grant Time: %u\r\n", metrics.mAvgRxRequestToGrantTime);
+        mServer->OutputFormat("    Grant None: %u\r\n", metrics.mNumRxGrantNone);
+    }
+    else if (strcmp(argv[0], "enable") == 0)
```
Comment (bukepo) [inline src/cli/cli.cpp]: How about the following syntax:
```bash
* coex enable - enable coex
* coex disable - disable coex
* coex metrics - output the metrics
* coex - output the state
```

File: examples/platforms/posix/radio.c (hunk 1 above)
Comment (zhanglongxia) [reply examples/platforms/posix/radio.c]: Done

File: src/cli/cli.cpp (hunk 2 above)
Comment (zhanglongxia) [reply src/cli/cli.cpp]: Done

File: examples/platforms/posix/radio.c [hunk 3]
```diff
@@ -1093,7 +1097,30 @@ int8_t otPlatRadioGetReceiveSensitivity(otInstance *aInstance)
     return POSIX_RECEIVE_SENSITIVITY;
 }
 
-#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE
+#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE
+otError otPlatRadioCoexEnable(otInstance *aInstance)
```
Comment (bukepo) [inline examples/platforms/posix/radio.c]: see #4159

File: src/ncp/spinel.h [hunk 4]
```diff
@@ -1664,6 +1664,13 @@ typedef enum
      */
     SPINEL_PROP_RADIO_COEX_METRICS = SPINEL_PROP_PHY_EXT__BEGIN + 12,
 
+    /// Radio Coex Enable
+    /** Format: `b`
+     *
+     * Indicates if radio coex is enabled or disabled. Set to true to enable radio coex.
+     */
+    SPINEL_PROP_RADIO_COEX_ENABLE = SPINEL_PROP_PHY_EXT__BEGIN + 13,
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/ncp/spinel.h]: - The source code of the [function](https://github.com/openthread/wpantund/blob/master/src/ncp-spinel/SpinelNCPInstance.cpp#L1863) shows that some spinel properties need to add related spinel capability and some don't need. Do we have a principle that defines which kind of spinel property should add related spinel capability?

-  If we add the spinel capability and the coex is disabled, the command  `wpanctl getprop NCP:Coex:Enable` shows `Capability SPINEL_CAP_RADIO_COEX (required for "NCP:Coex:Enable") is not supported by NCP`. If we don't add spinel capability  and the coex is disabled, the command  `wpanctl getprop NCP:Coex:Enable` shows `NCP:Coex:Enable: <empty> (16)`. From user aspect, we know that we can't get this property from the output info.  So what benefit we can get from adding the spinel capability?

[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/ncp/spinel.h]: @abtink Thanks for your detailed explanation about the spinel capability. I have updated the code to add the Coex spinel capability.

File: src/posix/platform/radio_spinel.hpp [hunk 5]
```diff
@@ -197,7 +197,29 @@ class RadioSpinel : public HdlcInterface::Callbacks
...
+
+    /**
+     * Check whether radio coex is enabled or not.
+     *
+     * @param[in] aInstance  The OpenThread instance structure.
+     *
+     * @returns TRUE if the radio coex is enabled, FALSE otherwise.
+     *
+     */
+    bool IsCoexEnabled(void);
```
Comment (jwhui) [inline src/posix/platform/radio_spinel.hpp]: Declare as `const`.

File: src/posix/platform/radio_spinel.cpp [hunk 6]
```diff
@@ -1670,10 +1684,22 @@ int8_t otPlatRadioGetReceiveSensitivity(otInstance *aInstance)
     return sRadioSpinel.GetReceiveSensitivity();
 }
 
-#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE
+#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE
+otError otPlatRadioSetCoexEnabled(otInstance *aInstance, bool aEnabled)
+{
+    assert(aInstance != NULL);
```
Comment (jwhui) [inline src/posix/platform/radio_spinel.cpp]: Remove `assert` to be consistent with rest of this file?

File: src/posix/platform/radio_spinel.cpp [hunk 7]
```diff
@@ -1670,10 +1684,22 @@ int8_t otPlatRadioGetReceiveSensitivity(otInstance *aInstance)
...
+    return sRadioSpinel.SetCoexEnabled(aEnabled);
+}
+
+bool otPlatRadioIsCoexEnabled(otInstance *aInstance)
+{
+    assert(aInstance != NULL);
```
Comment (jwhui) [inline src/posix/platform/radio_spinel.cpp]: Remove `assert` to be consistent with rest of this file?

File: src/posix/platform/radio_spinel.cpp [hunk 8]
```diff
@@ -1670,10 +1684,22 @@ int8_t otPlatRadioGetReceiveSensitivity(otInstance *aInstance)
...
+    return sRadioSpinel.IsCoexEnabled();
+}
+
 otError otPlatRadioGetCoexMetrics(otInstance *aInstance, otRadioCoexMetrics *aCoexMetrics)
 {
-    OT_UNUSED_VARIABLE(aInstance);
+    assert(aInstance != NULL);
```
Comment (jwhui) [inline src/posix/platform/radio_spinel.cpp]: Remove `assert` to be consistent with rest of this file?

File: src/posix/platform/radio_spinel.hpp (hunk 5 above)
Comment (zhanglongxia) [reply src/posix/platform/radio_spinel.hpp]: Function `IsCoexEnabled()` calls the function `Get()` to get the state of Coex. The function `Get()` is not a const function. So I can't declare it as `const`.

File: src/posix/platform/radio_spinel.cpp (hunk 6 above)
Comment (zhanglongxia) [reply src/posix/platform/radio_spinel.cpp]: Updated

File: src/posix/platform/radio_spinel.cpp (hunk 7 above)
Comment (zhanglongxia) [reply src/posix/platform/radio_spinel.cpp]: Updated

File: src/posix/platform/radio_spinel.cpp (hunk 8 above)
Comment (zhanglongxia) [reply src/posix/platform/radio_spinel.cpp]: Updated

**Pull request merged**